parlai/core/torch_generator_agent.py

```diff
@@ -995,3 +1023,60 @@ def select_paths(self, logprobs, prior_scores):
         tok_ids = best_idxs % voc_size
 
         return (hyp_ids, tok_ids, best_scores)
+
+
+class TopKSampling(TreeSearch):
+    """
+    Top-K sampling (Fan et al., 2018).
+
+    Samples from a truncated distribution where only the most probable K words
+    are considered at each time.
+
+    Typical values of k are 2, 10, 50.
+
+    See https://arxiv.org/abs/1805.04833 for details.
+    """
+
+    def __init__(self, k, *args, **kwargs):
+        self.k = k
+        super().__init__(*args, **kwargs)
+
+    def select_paths(self, logprobs, prior_scores):
+        topk = logprobs.topk(self.k, dim=-1)
+        probs = torch.softmax(topk.values, dim=-1)
+        choices = torch.multinomial(probs, 1)[:, 0]
+        hyp_ids = torch.arange(logprobs.size(0)).to(logprobs.device)
+        tok_ids = topk.indices[hyp_ids, choices]
+        scores = topk.values[hyp_ids, choices]
+        best_scores = prior_scores.expand_as(scores) + scores
+        return (hyp_ids, tok_ids, best_scores)
+
+
+class NucleusSampling(TreeSearch):
+    """
+    Nucelus, aka top-p sampling (Holtzman et al., 2019).
+
+    Samples from a truncated distribution which covers a fixed CDF proportion
+    of the original distribution.
+
+    Typical values of p are 0.3 and 0.9.
+
+    See https://arxiv.org/abs/1904.09751 for details.
+    """
+
+    def __init__(self, p, *args, **kwargs):
+        self.p = p
+        super().__init__(*args, **kwargs)
+
+    def select_paths(self, logprobs, prior_scores):
+        probs = torch.softmax(logprobs, dim=-1)
+        sprobs, sinds = probs.sort(dim=-1, descending=True)
+        mask = (sprobs.cumsum(dim=-1) - sprobs[:, :1]) > self.p
+        sprobs[mask] = 0
+        sprobs.div_(sprobs.sum(dim=-1).unsqueeze(1))
+        choices = torch.multinomial(sprobs, 1)[:, 0]
+        hyp_ids = torch.arange(logprobs.size(0)).to(logprobs.device)
+        tok_ids = sinds[hyp_ids, choices]
+        scores = sprobs[hyp_ids, choices].log()
```

Comment:
question, why do you take the log here?

Reply:
Because `best_scores` keeps track of the probability of the full utterance. It's much better to add probabilities in logspace than to multiply them in linspace.

Reply:
then why don't you do the same thing above in top k?

Reply:
top-k never leaves logspace. The top-p had to be converted to linspace to do the truncation. (Can't compute the CDF in logspace...)

Reply:
oh i see now, thanks

Reply:
I’ll leave a comment